File: app/src/main/java/com/practice/todo/util/LinearLayoutListHelper.java
package com.practice.todo.util;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;

import java.util.ArrayList;
import java.util.List;

public class LinearLayoutListHelper {
    public static void setAdapter(LinearLayout layout, LinearAdapter adapter) {
        layout.removeAllViews();
        for (int i = 0; i < adapter.getItemCount(); i++) {
            ViewHolder viewHolder = adapter.onCreateViewHolder(layout, adapter.getItemType(i));
            adapter.onBindViewHolder(viewHolder, i);
            layout.addView(viewHolder.itemView);
        }
    }

    public static abstract class LinearAdapter<VH extends ViewHolder> {

        protected Context mContext;
        protected LayoutInflater mInflater;

        public LinearAdapter(Context mContext) {
            this.mContext = mContext;
            mInflater = LayoutInflater.from(mContext);
        }

        public abstract VH onCreateViewHolder(LinearLayout parent, int type);

        public abstract void onBindViewHolder(VH holder, int position);

        public abstract int getItemCount();

        public int getItemType(int position) {
            return -1;
        }
    }

    public static abstract class SimpleLinearAdapter<VH extends ViewHolder, T> extends LinearAdapter<VH> {
        private final LinearLayout mParent;
        protected List<T> mDataSet = new ArrayList<>();

        public SimpleLinearAdapter(Context context, LinearLayout parent) {
            super(context);
            mParent = parent;
        }

        @Override
        public abstract VH onCreateViewHolder(LinearLayout parent, int type);

        @Override
        public void onBindViewHolder(VH holder, int position) {
            if (!mDataSet.isEmpty()) {
                onSimpleBindView(holder, mDataSet.get(position), position);
            }
        }

        protected abstract void onSimpleBindView(VH holder, T t, int position);

        @Override
        public int getItemCount() {
            return mDataSet.size();
        }

        public void refreshData(List<T> list) {
            mDataSet.clear();
            mDataSet.addAll(list);
            setAdapter(mParent, this);
        }
    }

    public static class ViewHolder {
        public View itemView;

        public ViewHolder(View itemView) {
            this.itemView = itemView;
        }
    }
}

File: app/src/main/java/com/practice/todo/util/LinearLayoutListHelper.kt
package com.practice.todo.util

import android.view.View
import android.widget.LinearLayout


/**
 * Created by suikajy on 2019.10.2
 */

class LinearLayoutListHelper {

    companion object {
        fun <VH : ViewHolder> setAdapter(layout: LinearLayout, adapter: LinearAdapter<VH>) {
            layout.removeAllViews()
            for (i in 0 until adapter.itemCount) {
                val viewHolder = adapter.onCreateViewHolder(layout, adapter.getItemType(i))
                adapter.onBindViewHolder(viewHolder, i)
                layout.addView(viewHolder.itemView)
            }
        }
    }

    abstract class LinearAdapter<VH : ViewHolder> {
        abstract val itemCount: Int

        abstract fun onCreateViewHolder(parent: LinearLayout, type: Int): VH

        abstract fun onBindViewHolder(holder: VH, position: Int)

        fun getItemType(position: Int) = -1
    }

    abstract class SimpleLinearAdapter<VH : ViewHolder, T>(
        private val mParent: LinearLayout
    ) : LinearAdapter<VH>() {
        protected var mDataSet: MutableList<T> = ArrayList()

        override val itemCount: Int
            get() = mDataSet.size

        abstract override fun onCreateViewHolder(parent: LinearLayout, type: Int): VH

        override fun onBindViewHolder(holder: VH, position: Int) {
            if (mDataSet.isNotEmpty()) {
                onSimpleBindView(holder, mDataSet[position], position)
            }
        }

        protected abstract fun onSimpleBindView(holder: VH, item: T, position: Int)

        fun refreshData(list: List<T>) {
            mDataSet.clear()
            mDataSet.addAll(list)
            setAdapter(mParent, this)
        }

        fun clear() {
            mDataSet.clear()
            setAdapter(mParent, this)
        }

        fun addItem(element: T) {
            val elePosition = mDataSet.size
            mDataSet.add(element)
            val viewHolder = onCreateViewHolder(mParent, getItemType(elePosition))
            onBindViewHolder(viewHolder, elePosition)
            mParent.addView(viewHolder.itemView)
        }
    }

    open class ViewHolder(val itemView: View)

}